compiler/internal/cueutil/build.go

```diff
@@ -0,0 +1,268 @@
+	ctx := cuecontext.New()
+	values, err := ctx.BuildInstances(pkgs)
+	if err != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+	if len(values) == 0 {
+		return cue.Value{}, eerror.New("config", "no values generated from config", nil)
+	}
+
+	// Unify all returned values into a single value
+	rtnValue := values[0]
+	for _, value := range values {
+		if value.Err() != nil {
+			return cue.Value{}, toDescriptiveError(value.Err())
+		}
+		rtnValue = rtnValue.Unify(value)
+	}
+
+	// Check the unified value for errors
+	if rtnValue.Err() != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+
+	// Validate the unified value is concrete
+	if err := rtnValue.Validate(cue.Concrete(true)); err != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+
+	return rtnValue, nil
+}
+
+// allFilesUnder returns all files under the given path in the given filesystem.
+func allFilesUnder(filesys fs.FS, path string) ([]string, error) {
+	var files []string
+	err := fs.WalkDir(filesys, path, func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+// toDescriptiveError takes a CUE error list and expands it to show each error (unless there is more than 10)
+func toDescriptiveError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	errs := cueerrors.Errors(err)
+	if len(errs) == 1 {
+		return errs[0]
+	}
+
+	var str strings.Builder
+	str.WriteString("the following errors where detected:\n")
+	for i, err2 := range errs {
+		str.WriteString(fmt.Sprintf("\t- %s\n", err2))
+
+		if i >= 10 {
+			str.WriteString("\t- ... (too many errors to show)")
+			break
+		}
+	}
+
+	return errors.New(str.String())
+}
+
+// writeFSToTmpPath writes the contents of the given filesystem to a temporary directory on the local filesystem.
+func writeFSToTmpPath(filesys fs.FS) (string, error) {
+	// Work out of a temporary directory
+	tmpPath, err := os.MkdirTemp("", "encr-cfg-")
+	if err != nil {
+		return "", eerror.Wrap(err, "config", "unable to create temporary directory", nil)
+	}
+
+	// Copy the files into the temporary directory
+	err = fs.WalkDir(filesys, ".", func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() {
+			// Open the source file from our filesystem
+			srcFile, err := filesys.Open(path)
+			if err != nil {
+				return err
+			}
+
+			dstFile, err := os.OpenFile(filepath.Join(tmpPath, path), os.O_CREATE|os.O_WRONLY, 0644)
+			if err != nil {
+				return err
+			}
+
+			_, err = io.Copy(dstFile, srcFile)
+			if err != nil {
+				return err
+			}
+		} else {
+			if err := os.Mkdir(filepath.Join(tmpPath, path), 0755); err != nil && !errors.Is(err, os.ErrExist) {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		_ = os.RemoveAll(tmpPath) // clear the temporary directory incase of error
```

Comment:
See comment earlier in the PR: better to move the temp dir creation up a level and remove this from here